### PR TITLE
Fixed #29172 -- Fixed crash with Window expression in a subquery.

### PR DESCRIPTION
**Pull request by tricoder42**
[Ticket](https://code.djangoproject.com/ticket/29172#ticket)

### CODE REVIEW

**Comment by carltongibson** [comment]
Hi @tricoder42. There were some failures here (which didn't reproduce) so I've squashed and re-pushed to see if that fixes the build.

**Comment by tricoder42** [comment]
@carltongibson I think there's a problem with daylight saving 🕑 

Looking at the [failed test](https://djangoci.com/job/pull-requests-trusty/database=postgres,label=trusty-pr,python=python3.6/17244/testReport/junit/basic.tests/ModelRefreshTests/test_refresh/), the datetime is 11th March (`datetime.datetime(2018, 3, 11, 3, 5, 0, 317139)`) when daylight saving in [US starts](https://www.timeanddate.com/time/change/usa).

File: tests/expressions_window/tests.py [hunk 1]
```diff
@@ -722,6 +722,21 @@ def test_invalid_type_start_row_range(self):
                 frame=RowRange(start='a'),
             )))
 
+    @skipUnlessDBFeature('supports_over_clause')
+    def test_window_expression_within_subquery(self):
+        subquery_qs = Employee.objects.annotate(
+            highest=Window(FirstValue('id'), partition_by=F('department'), order_by=F('salary').desc())
+        ).values('highest')
+        highest_salary = Employee.objects.filter(pk__in=Subquery(subquery_qs))
```
Comment (carltongibson) [inline tests/expressions_window/tests.py]: Do we need to use `Subquery` here? (If I remove it the test still functions — i.e. fails without patch, passes with it.)

Comment (tricoder42) [reply tests/expressions_window/tests.py]: You're right! The `Subquery` is implicit in this case and not required.

```sql
SELECT
"expressions_window_employee"."id", "expressions_window_employee"."name", "expressions_window_employee"."salary", "expressions_window_employee"."department", "expressions_window_employee"."hire_date" 
FROM "expressions_window_employee" 
WHERE "expressions_window_employee"."id" IN (
    SELECT FIRST_VALUE(U0."id") OVER (PARTITION BY U0."department" ORDER BY U0."salary" DESC
) AS "highest" FROM "expressions_window_employee" U0)
```

Comment (tricoder42) [reply tests/expressions_window/tests.py]: You're right! The `Subquery` is implicit in this case and not required.

```sql
SELECT
"expressions_window_employee"."id", "expressions_window_employee"."name", "expressions_window_employee"."salary", "expressions_window_employee"."department", "expressions_window_employee"."hire_date" 
FROM "expressions_window_employee" 
WHERE "expressions_window_employee"."id" IN (
    SELECT FIRST_VALUE(U0."id") OVER (PARTITION BY U0."department" ORDER BY U0."salary" DESC
) AS "highest" FROM "expressions_window_employee" U0)
```

**Comment by tricoder42** [comment]
@carltongibson Regarding the failing test case: It really fails only on 1st March from 2am to 3am eastern time. I was lucky enough to run it exactly in this window ;)

Simple patch to reproduce the bug:

```diff
diff --git a/tests/basic/tests.py b/tests/basic/tests.py
index 1a5e95e9d6..84ef27357a 100644
--- a/tests/basic/tests.py
+++ b/tests/basic/tests.py
@@ -639,8 +639,9 @@ class SelectOnSaveTests(TestCase):
 class ModelRefreshTests(TestCase):

     def test_refresh(self):
-        a = Article.objects.create(pub_date=datetime.now())
-        Article.objects.create(pub_date=datetime.now())
+        now = datetime(2018, 3, 1, 2, 5, 0, 317139)
+        a = Article.objects.create(pub_date=now)
+        Article.objects.create(pub_date=now)
         Article.objects.filter(pk=a.pk).update(headline='new headline')
         with self.assertNumQueries(1):
             a.refresh_from_db()
```


However, this is for another PR 😄

**Pull request merged**